cloud/pkg/devicecontroller/controller/downstream.go

```diff
@@ -121,28 +124,70 @@ func (dc *DownstreamController) syncDevice() {
 
 // deviceAdded creates a device, adds in deviceManagers map, send a message to edge node if node selector is present.
 func (dc *DownstreamController) deviceAdded(device *v1beta1.Device) {
-	dc.deviceManager.Device.Store(device.Name, device)
+	// Nothing to do when device added, only add in UndeployedDevice map
+	dc.deviceManager.UndeployedDevice.Store(device.Name, device)
+	// Determine which node the Device should deploy
+	nodeID := ""
 	if device.Spec.NodeName != "" {
-		edgeDevice := createDevice(device)
-		msg := model.NewMessage("")
-
-		resource, err := messagelayer.BuildResourceForDevice(device.Spec.NodeName, "membership", "")
+		// If device.Spec.NodeName is present, directly define the node where the device should be deployed
+		nodeID = device.Spec.NodeName
+	} else {
+		// If device.Spec.NodeName is not present
+		// Find the node where device.MapperRef corresponds to mapper deployed
+		value, ok := dc.mapperManager.Mapper2NodeMap.Load(device.Spec.MapperRef.Name)
```

Comment:
Does users have to define MapperRef in device. yaml? If the user does not define MapperRef field, a null pointer error will occur here